cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -723,53 +723,85 @@ func fixNodeGroupSize(context *context.AutoscalingContext, clusterStateRegistry
 }
 
 // Removes unregistered nodes if needed. Returns true if anything was removed and error if such occurred.
-func removeOldUnregisteredNodes(unregisteredNodes []clusterstate.UnregisteredNode, context *context.AutoscalingContext,
+func (a *StaticAutoscaler) removeOldUnregisteredNodes(unregisteredNodes []clusterstate.UnregisteredNode, context *context.AutoscalingContext,
 	csr *clusterstate.ClusterStateRegistry, currentTime time.Time, logRecorder *utils.LogEventRecorder) (bool, error) {
-	removedAny := false
-	for _, unregisteredNode := range unregisteredNodes {
-		nodeGroup, err := context.CloudProvider.NodeGroupForNode(unregisteredNode.Node)
+
+	nodeGroups := a.nodeGroupsById()
+	nodesToBeDeletedByNodeGroupId := make(map[string][]clusterstate.UnregisteredNode)
+	for _, un := range unregisteredNodes {
+		nodeGroup, err := a.CloudProvider.NodeGroupForNode(un.Node)
 		if err != nil {
-			klog.Warningf("Failed to get node group for %s: %v", unregisteredNode.Node.Name, err)
-			return removedAny, err
-		}
-		if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
-			klog.Warningf("No node group for node %s, skipping", unregisteredNode.Node.Name)
+			id := "<nil>"
+			if un.Node != nil {
+				id = un.Node.Spec.ProviderID
+			}
+			klog.Warningf("Cannot determine nodeGroup for node %v; %v", id, err)
 			continue
 		}
 
 		maxNodeProvisionTime, err := csr.MaxNodeProvisionTime(nodeGroup)
 		if err != nil {
-			return removedAny, fmt.Errorf("failed to retrieve maxNodeProvisionTime for node %s in nodeGroup %s", unregisteredNode.Node.Name, nodeGroup.Id())
+			return false, fmt.Errorf("failed to retrieve maxNodeProvisionTime for node %s in nodeGroup %s", un.Node.Name, nodeGroup.Id())
 		}
 
-		if unregisteredNode.UnregisteredSince.Add(maxNodeProvisionTime).Before(currentTime) {
-			klog.V(0).Infof("Removing unregistered node %v", unregisteredNode.Node.Name)
-			size, err := nodeGroup.TargetSize()
-			if err != nil {
-				klog.Warningf("Failed to get node group size; unregisteredNode=%v; nodeGroup=%v; err=%v", unregisteredNode.Node.Name, nodeGroup.Id(), err)
-				continue
+		if un.UnregisteredSince.Add(maxNodeProvisionTime).Before(currentTime) {
+			if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
+				a.clusterStateRegistry.RefreshCloudProviderNodeInstancesCache()
+				return false, fmt.Errorf("node %s has no known nodegroup", un.Node.GetName())
```

Comment:
Right, I added it based on the other one... I agree, this might have been a mistake - better to leave it closer to the original shape. I'm not sure why the logic differs. Reverted (same for the change above).